### PR TITLE
EVAKA-4240: display attendances on weekly reservations view

### PR DESCRIPTION
**Pull request by Joosakur**
#### Summary

- Improve api to support children changing groups during week
- Include attendances and absences in the response
- display attendances
- display service time defaults
- displaying absences (more absence types to be handled later)

### CODE REVIEW

File: service/src/main/kotlin/fi/espoo/evaka/reservations/AttendanceReservationController.kt [hunk 1]
```diff
@@ -78,120 +89,163 @@ private fun Database.Read.getUnitOperationalDays(unitId: DaycareId, dateRange: F
...
+
     data class GroupAttendanceReservations(
         val group: String,
         val children: List<ChildReservations>
     )
 
     data class ChildReservations(
         val child: Child,
-        val reservations: Map<LocalDate, Reservation>
+        val dailyData: List<DailyChildData>
```
Comment (jaahon) [inline service/src/main/kotlin/fi/espoo/evaka/reservations/AttendanceReservationController.kt:190]: I don't really understand why you converted this into a list.

File: service/src/main/kotlin/fi/espoo/evaka/reservations/AttendanceReservationController.kt [hunk 2]
```diff
@@ -78,120 +89,163 @@ private fun Database.Read.getUnitOperationalDays(unitId: DaycareId, dateRange: F
...
+        att.departed AS attendance_end,
+        ab.absence_type
+    FROM generate_series(:start, :end, '1 day') t
+    JOIN placement pl ON pl.unit_id = :unitId AND daterange(pl.start_date, pl.end_date, '[]') @> t::date
+    JOIN person p ON p.id = pl.child_id
+    LEFT JOIN daycare_group_placement dgp on dgp.daycare_placement_id = pl.id AND daterange(dgp.start_date, dgp.end_date, '[]') @> t::date
+    LEFT JOIN daycare_group dg ON dg.id = dgp.daycare_group_id
+    LEFT JOIN attendance_reservation res ON res.child_id = p.id AND res.start_date = t::date
+    LEFT JOIN child_attendance att ON att.child_id = p.id AND (att.arrived AT TIME ZONE 'Europe/Helsinki')::date = t::date
+    LEFT JOIN absence ab ON ab.child_id = p.id AND ab.date = t::date
```
Comment (jaahon) [inline service/src/main/kotlin/fi/espoo/evaka/reservations/AttendanceReservationController.kt:113]: This join means that multiple rows are returned when a child has multiple absences during the same day.

File: service/src/main/kotlin/fi/espoo/evaka/reservations/AttendanceReservationController.kt [hunk 3]
```diff
@@ -78,120 +89,163 @@ private fun Database.Read.getUnitOperationalDays(unitId: DaycareId, dateRange: F
...
+                            reservation = rowByDateAndChild.reservation?.let {
+                                UnitAttendanceReservations.ReservationTimes(
+                                    startTime = it.startTime,
+                                    endTime = it.endTime
+                                )
+                            },
+                            attendance = rowByDateAndChild.attendance?.let {
+                                UnitAttendanceReservations.AttendanceTimes(
+                                    startTime = it.start.toLocalTime().format(DateTimeFormatter.ofPattern("HH:mm")),
+                                    endTime = it.end?.toLocalTime()?.format(DateTimeFormatter.ofPattern("HH:mm")),
```
Comment (jaahon) [inline service/src/main/kotlin/fi/espoo/evaka/reservations/AttendanceReservationController.kt:156]: For consistency either do this formatting in the database query like with reservations or convert the reservations into local times in code.

File: service/src/main/kotlin/fi/espoo/evaka/reservations/AttendanceReservationController.kt [hunk 4]
```diff
@@ -78,120 +89,163 @@ private fun Database.Read.getUnitOperationalDays(unitId: DaycareId, dateRange: F
...
+
+    data class DailyChildData(
+        val date: LocalDate,
+        val reservation: ReservationTimes?,
+        val attendance: AttendanceTimes?,
+        val absence: Absence?
     )
 
-    data class Reservation(
+    data class ReservationTimes(
```
Comment (jaahon) [inline service/src/main/kotlin/fi/espoo/evaka/reservations/AttendanceReservationController.kt:200]: As it is this class is identical to `QueryRowReservation`. Doesn't seem like they are both needed. The same is true for `Absence` and `QueryRowAbsence`.

File: service/src/main/kotlin/fi/espoo/evaka/reservations/AttendanceReservationController.kt [hunk 5]
```diff
@@ -39,25 +46,29 @@ class AttendanceReservationController(private val ac: AccessControl) {
         if (to < from || from.plusMonths(1) < to) throw BadRequest("Invalid query dates")
         val dateRange = FiniteDateRange(from, to)
         return db.read { tx ->
+            val unitName = tx.getDaycare(unitId)?.name ?: throw NotFound("Unit $unitId not found")
             val operationalDays = tx.getUnitOperationalDays(unitId, dateRange)
             tx
-                .getUnitAttendanceReservations(unitId, dateRange)
+                .getAttendanceReservationData(unitId, dateRange)
+                .let { it }
```
Comment (jaahon) [inline service/src/main/kotlin/fi/espoo/evaka/reservations/AttendanceReservationController.kt:53]: Remove this.

File: service/src/main/kotlin/fi/espoo/evaka/dailyservicetimes/DailyServiceTimes.kt [hunk 6]
```diff
@@ -33,21 +33,27 @@ enum class DailyServiceTimesType {
 
 @JsonTypeInfo(use = JsonTypeInfo.Id.DEDUCTION, property = "type")
 sealed class DailyServiceTimes(
+    open val childId: UUID,
```
Comment (jaahon) [inline service/src/main/kotlin/fi/espoo/evaka/dailyservicetimes/DailyServiceTimes.kt:36]: I don't understand the motivation behind adding `childId` to this class. Seems like unnecessary coupling.

File: frontend/src/employee-frontend/components/unit/unit-reservations/ReservationsTable.tsx [hunk 7]
```diff
@@ -6,18 +6,113 @@ import React from 'react'
...
+  isRegular,
+  isVariableTime
+} from 'lib-common/api-types/child/common'
 
 interface Props {
   operationalDays: OperationalDay[]
   reservations: ChildReservations[]
 }
 
+function renderAbsence(type: AbsenceType) {
```
Comment (jaahon) [inline frontend/src/employee-frontend/components/unit/unit-reservations/ReservationsTable.tsx:31]: This function and `renderChildDay` could just be regular components.

File: frontend/src/employee-frontend/components/unit/unit-reservations/ReservationsTable.tsx [hunk 8]
```diff
@@ -6,18 +6,113 @@ import React from 'react'
...
+
+  if (day.isHoliday && !dailyData.reservation && !dailyData.attendance)
+    return null
+
+  if (dailyData.absence && !dailyData.attendance)
+    return renderAbsence(dailyData.absence.type)
+
+  const serviceTimes = childReservations.child.dailyServiceTimes
+  const serviceTimesAvailable =
+    serviceTimes != null && !isVariableTime(serviceTimes)
```
Comment (jaahon) [inline frontend/src/employee-frontend/components/unit/unit-reservations/ReservationsTable.tsx:74]: Use strict equality.

File: frontend/src/employee-frontend/components/unit/unit-reservations/ReservationsTable.tsx [hunk 9]
```diff
@@ -6,18 +6,113 @@ import React from 'react'
...
+        ) : serviceTimesAvailable && serviceTimeOfDay ? (
+          <>
+            <ReservationTime>{serviceTimeOfDay.start}*</ReservationTime>
+            <ReservationTime>{serviceTimeOfDay.end}*</ReservationTime>
+          </>
+        ) : serviceTimesAvailable && serviceTimeOfDay === null ? (
+          <ReservationTime>Vapaapäivä</ReservationTime>
+        ) : (
+          <ReservationTime>Ei varausta</ReservationTime>
+        )}
```
Comment (jaahon) [inline frontend/src/employee-frontend/components/unit/unit-reservations/ReservationsTable.tsx:110]: It's pretty hard to follow what's happening here.

File: service/src/main/kotlin/fi/espoo/evaka/dailyservicetimes/DailyServiceTimes.kt (hunk 6 above)
Comment (Joosakur) [reply service/src/main/kotlin/fi/espoo/evaka/dailyservicetimes/DailyServiceTimes.kt:36]: It was needed so that when querying service times for week view we know which row is for which child

File: service/src/main/kotlin/fi/espoo/evaka/reservations/AttendanceReservationController.kt (hunk 1 above)
Comment (Joosakur) [reply service/src/main/kotlin/fi/espoo/evaka/reservations/AttendanceReservationController.kt:190]: Hmm, it might not be necessary after all, I'll give it another try

File: service/src/main/kotlin/fi/espoo/evaka/reservations/AttendanceReservationController.kt (hunk 2 above)
Comment (Joosakur) [reply service/src/main/kotlin/fi/espoo/evaka/reservations/AttendanceReservationController.kt:113]: Right.. any suggestion on how to fix? I guess on DB level they may even have different absence types, but the UI spec does not support showing multiple reasons for absence per day.

Comment (jaahon) [reply service/src/main/kotlin/fi/espoo/evaka/reservations/AttendanceReservationController.kt:113]: I don't really know how it should be interpreted in case the absence types differ, but I would just take a random one with a lateral join for now.

File: service/src/main/kotlin/fi/espoo/evaka/dailyservicetimes/DailyServiceTimes.kt (hunk 6 above)
Comment (Joosakur) [reply service/src/main/kotlin/fi/espoo/evaka/dailyservicetimes/DailyServiceTimes.kt:36]: removed

File: frontend/src/employee-frontend/components/unit/unit-reservations/ReservationsTable.tsx (hunk 9 above)
Comment (Joosakur) [reply frontend/src/employee-frontend/components/unit/unit-reservations/ReservationsTable.tsx]: added comments, not sure how to make the code itself more readable

**Review by jaahon — APPROVED** [review]
